src/shared/btrfs-util.c

```diff
@@ -1360,12 +1360,36 @@ static int subvol_snapshot_children(
 
         strncpy(vol_args.name, subvolume, sizeof(vol_args.name)-1);
 
-        if (ioctl(new_fd, BTRFS_IOC_SNAP_CREATE_V2, &vol_args) < 0)
-                return -errno;
+        for (;;) {
+                if (ioctl(new_fd, BTRFS_IOC_SNAP_CREATE_V2, &vol_args) < 0)
+                        return -errno;
+
+                if (!FLAGS_SET(flags, BTRFS_SNAPSHOT_LOCK_BSD))
+                        break;
+
+                subvolume_fd = xopenat_lock(new_fd, subvolume,
+                                            O_RDONLY|O_NOCTTY|O_CLOEXEC|O_DIRECTORY|O_NOFOLLOW,
+                                            /* xopen_flags = */ 0,
+                                            /* mode = */ 0,
+                                            LOCK_BSD,
+                                            LOCK_EX);
+                if (subvolume_fd >= 0) {
+                        if (btrfs_is_subvol_fd(subvolume_fd))
+                                break;
+
+                        /* Not a subvolume? Someone must have deleted the subvolume from under us and
+                        * replaced it with a directory, let's remove the directory and try again. */
```

Comment:
any reason why you are not just failing in this case?

Reply:
hmm, xopenat_lock() does that only to deal we external lock files, i.e. files that are created just for the purpose of locking them, and are then deleted once the lock is released. if implemented trivially this might be racy because we cannot open + lock a file atomically, hence we have to open it, then lock it, and then check that no privous owener deleted it in the time between the open and the lock, and we have an fd to an unlinked inode...

also, note thta the loop in xopenat_lock() doesn't actually unlink anything.

But this doesn't apply to subvols at all, those are never sed as lockfiles... 

i think this stuff delete+retry stuff when snapshotting should be removed, it doesn't make any sense to me.